File: leetcode/Arrays/leetcode-0057/57.java
class Solution {
    public int[][] insert(int[][] intervals, int[] newInterval) {        
        List<int[]> list = new ArrayList<>();
        for (int[] interval : intervals) {
            if (newInterval != null && interval[0] >= newInterval[0]) {
                list.add(newInterval);
                newInterval = null;
            }
            list.add(interval);
        }
        if (newInterval != null) {
            list.add(newInterval);
        }

        List<int[]> res = new ArrayList<>();
        int[] tmp = list.get(0);
        for (int i = 1; i < list.size(); i++) {
            if (tmp[1] >= list.get(i)[0]) {
                tmp[1] = Math.max(tmp[1], list.get(i)[1]);
            } else {
                res.add(tmp);
                tmp = list.get(i);
            }
        }
        res.add(tmp);

        int[][] ret = new int[res.size()][];
        int j = 0;
        for (int[] r : res) {
            ret[j++] = r;
        }
        return ret;
    }
}

File: leetcode/Arrays/leetcode-0057/57.go
package arrays

func insert(intervals [][]int, newInterval []int) [][]int {
	if len(intervals) == 0 {
		intervals = append(intervals, newInterval)
	} else {
		i := 0
		for i < len(intervals) && intervals[i][0] < newInterval[0] {
			i++
		}
		if i < len(intervals) {
			intervals = append(intervals[:i+1], intervals[i:]...)
			intervals[i] = newInterval
		} else {
			intervals = append(intervals, newInterval)
		}
	}

	res := [][]int{}
	tmp := intervals[0]
	for i := 1; i < len(intervals); i++ {
		if intervals[i][0] <= tmp[1] {
			if intervals[i][1] > tmp[1] {
				tmp[1] = intervals[i][1]
			}
		} else {
			res = append(res, tmp)
			tmp = intervals[i]
		}
	}
	res = append(res, tmp)
	return res
}
